### PR TITLE
raftstore: fix merge panic when multiple snapshots

### PR DESCRIPTION
**Pull request by Connor1996**
## What have you changed? (mandatory)
When that a follower is isolated, then in some extreme case it may receive two different snapshots, one before merge and one after merge. Once it applies the first snapshot and then receives the second snapshot which updates `pending_cross_snap` and lets source peer destroy itself.  After applying the first snapshot, the follower is able to append log and do merge itself. But now the source peer is destroyed so merge will panic.

The essential reason is that snapshot is unreliable, we can not make a decision based on the snapshot. Even we receive a snapshot, it can't ensure that follower cannot catch the process by appending logs. The snapshot may be generated in some case, like leader transfer or network issue.

So we can decide whether source peer to destroy by target peer when there is a new snapshot which ensures that once the source peer decides to destroy itself based on one snapshot, and that snapshot must be applied. 

## What are the type of the changes? (mandatory)
- Bug fix (non-breaking change which fixes an issue)

## How has this PR been tested? (mandatory)
unit-test
integration-test

## Does this PR affect documentation (docs) update? (mandatory)
No

## Does this PR affect tidb-ansible update? (mandatory)
No

### CODE REVIEW

**Comment by Connor1996** [comment]
PTAL @BusyJay @hicqu @overvenus

File: components/test_raftstore/src/transport_simulate.rs [hunk 1]
```diff
@@ -592,6 +593,7 @@ impl Filter<StoreMsg> for LeadingDuplicatedSnapshotFilter {
         let mut stale = self.stale.load(Ordering::Relaxed);
         if stale {
             if last_msg.is_some() {
+                thread::sleep(Duration::from_millis(100));
```
Comment (zhangjinpeng1987) [inline components/test_raftstore/src/transport_simulate.rs]: Please add comments for this sleeping.

File: src/raftstore/store/fsm/peer.rs [hunk 2]
```diff
@@ -840,34 +840,46 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
     fn need_gc_merge(&mut self, msg: &RaftMessage) -> Result<bool> {
         let merge_target = msg.get_merge_target();
         let target_region_id = merge_target.get_id();
+        debug!(
+            "{} receive merge targets {:?}",
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/fsm/peer.rs]: Better to use new format logging.

**Comment by zhangjinpeng1987** [comment]
@BusyJay @overvenus PTAL

File: src/raftstore/store/fsm/peer.rs [hunk 3]
```diff
@@ -1020,25 +1009,54 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
...
+            .filter(|r| r.get_id() != region_id)
+        {
+            info!(
+                "{} region overlapped {:?}, {:?}",
+                self.fsm.peer.tag, exist_region, snap_region
+            );
+            // In some extreme case, it may happen that a new snapshot is received whereas a snapshot is still in applying
+            // if the snapshot under applying is generated before merge and the new snapshot is generated after merge,
+            // here may cause source peer destroyed improperly. So just drop message here if peer is applying snapshot.
+            if !self.fsm.peer.is_applying_snapshot() && !self.fsm.peer.has_pending_snapshot() {
```
Comment (BusyJay) [inline src/raftstore/store/fsm/peer.rs]: The condition check can be moved out of the for loop.

File: src/raftstore/store/fsm/peer.rs [hunk 4]
```diff
@@ -1020,25 +1009,54 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
...
+                if maybe_destroy_source(
```
Comment (BusyJay) [inline src/raftstore/store/fsm/peer.rs]: What if there is still a CommitMerge pending in apply thread?

File: src/raftstore/store/fsm/peer.rs [hunk 5]
```diff
@@ -2469,6 +2528,31 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
     }
 }
 
+/// Checks merge target, returns whether the source peer should be destroyed.
```
Comment (overvenus) [inline src/raftstore/store/fsm/peer.rs]: Could you add conditions when the function returns true?

File: src/raftstore/store/fsm/store.rs [hunk 6]
```diff
@@ -1325,35 +1337,56 @@ impl<'a, T: Transport, C: PdClient> StoreFsmDelegate<'a, T, C> {
...
+            ) {
+                if maybe_destroy_source(
+                    meta,
+                    region_id,
+                    exist_region.get_id(),
+                    msg.get_region_epoch().to_owned(),
                 ) {
-                    meta.pending_cross_snap
-                        .insert(region_id, msg.get_region_epoch().to_owned());
+                    regions_to_destroy.push(exist_region.get_id());
```
Comment (overvenus) [inline src/raftstore/store/fsm/store.rs]: Why not send `MergeResult` here?

File: src/raftstore/store/fsm/peer.rs [hunk 7]
```diff
@@ -840,34 +840,60 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
...
+                );
+            }
+        }
+        if let Some(epoch) = meta
+            .regions
+            .get(&target_region_id)
+            .map(|r| r.get_region_epoch())
+        {
+            // In the case that the source peer's range isn't overlapped with target's anymore:
+            //     | region 1 | region 2 | region 3 |
```
[the remark this answers is not in the document]
Comment (Connor1996) [reply src/raftstore/store/fsm/peer.rs]: why, it doesn't matter.

File: src/raftstore/store/fsm/store.rs (hunk 6 above)
Comment (Connor1996) [reply src/raftstore/store/fsm/store.rs]: there may be several overlapping regions, and other regions may not pass the check. So we only destroy them only if all the overlapping regions pass the check.

File: src/raftstore/store/fsm/peer.rs [hunk 8]
```diff
@@ -840,34 +840,60 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
...
+        // is the state of target peer at the time when source peer is merged. So here we record the
+        // merge target epoch version to let the target peer on this store to decide whether to
+        // destroy the source peer.
+        let mut meta = self.ctx.store_meta.lock().unwrap();
+        meta.targets_map.insert(self.region_id(), target_region_id);
+        let v = meta
+            .pending_merge_targets
+            .entry(target_region_id)
+            .or_default();
+        if let Some(epoch) = (*v).insert(self.region_id(), merge_target.get_region_epoch().clone())
```
[the remark this answers is not in the document]
Comment (Connor1996) [reply src/raftstore/store/fsm/peer.rs]: I don't understand. It does use `Entry` now.

File: components/test_raftstore/src/transport_simulate.rs [hunk 9]
```diff
@@ -592,6 +593,8 @@ impl Filter<StoreMsg> for LeadingDuplicatedSnapshotFilter {
         let mut stale = self.stale.load(Ordering::Relaxed);
         if stale {
             if last_msg.is_some() {
+                // To make sure the messages will not handled in one raftstore batch.
+                thread::sleep(Duration::from_millis(100));
```
[the remark this answers is not in the document]
Comment (Connor1996) [reply components/test_raftstore/src/transport_simulate.rs]: don't think necessary

File: src/raftstore/store/fsm/peer.rs [hunk 10]
```diff
@@ -1020,25 +1023,62 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
...
+        // CommitMerge may panic because source is already destroyed, so just drop the message:
+        // 1. A new snapshot is received whereas a snapshot is still in applying, and the snapshot
+        // under applying is generated before merge and the new snapshot is generated after merge.
+        // After the applying snapshot is finished, the log may able to catch up and so a
+        // CommitMerge will be applied.
+        // 2. There is a CommitMerge pending in apply thread.
+        let ready = !self.fsm.peer.is_applying_snapshot()
+            && !self.fsm.peer.has_pending_snapshot()
+            && self.fsm.peer.ready_to_handle_pending_snap();
+        for exist_region in meta
```
[the remark this answers is not in the document]
Comment (Connor1996) [reply src/raftstore/store/fsm/peer.rs]: it is clear enough, seems unnecessary.

[the remark this answers is not in the document]
Comment (overvenus) [reply src/raftstore/store/fsm/peer.rs]: Okay, to be honest it takes me 5 minutes to understand what is does.

File: components/test_raftstore/src/transport_simulate.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (overvenus) [reply components/test_raftstore/src/transport_simulate.rs]: What if raftstore runs slowly? Will they be batched?

File: src/raftstore/store/fsm/peer.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (overvenus) [reply src/raftstore/store/fsm/peer.rs]: I mean the `insert`.

File: src/raftstore/store/fsm/peer.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (overvenus) [reply src/raftstore/store/fsm/peer.rs]: I think it does not consistent with `|        region 4       |region 1|`, the split direction is different.

File: src/raftstore/store/fsm/store.rs [hunk 11]
```diff
@@ -1325,35 +1337,56 @@ impl<'a, T: Transport, C: PdClient> StoreFsmDelegate<'a, T, C> {
...
+                    continue;
                 }
-
-                return Ok(false);
             }
+            self.ctx.raft_metrics.message_dropped.region_overlap += 1;
```
Comment (BusyJay) [inline src/raftstore/store/fsm/store.rs]: Why move it here?

Comment (Connor1996) [reply src/raftstore/store/fsm/store.rs]: only dropped it should be count

**Comment by Connor1996** [comment]
PTAL again @BusyJay @overvenus

**Comment by Connor1996** [comment]
Some tiny change, PTAL again @BusyJay @overvenus

**Pull request merged**